File: ios/Sources/HovrDevicePermissions/PermissionActionResolver.swift
import Foundation

public enum PermissionCoordinatorAction: Equatable {
    case dismiss
    case requestSystem
    case showRequired
    case showDeviceLocationDisabled
    case showDeviceLocationRestricted
}

public enum PermissionActionResolver {
    public static func resolveLocationAction(
        issue: PermissionAccessIssue
    ) -> PermissionCoordinatorAction {
        switch issue {
        case .granted:
            return .dismiss
        case .deviceLocationDisabled:
            return .showDeviceLocationDisabled
        case .deviceLocationRestricted:
            return .showDeviceLocationRestricted
        case .notDetermined:
            return .requestSystem
        case .appLocationDenied:
            return .showRequired
        case .appNotificationDenied:
            return .dismiss
        }
    }

    public static func resolveNotificationAction(
        issue: PermissionAccessIssue
    ) -> PermissionCoordinatorAction {
        switch issue {
        case .granted:
            return .dismiss
        case .notDetermined:
            return .requestSystem
        case .appNotificationDenied:
            return .showRequired
        case .deviceLocationDisabled,
             .deviceLocationRestricted,
             .appLocationDenied:
            return .dismiss
        }
    }
}

File: android/hovr_device_permissions/src/main/kotlin/com/hovr/devicepermissions/PermissionActionResolver.kt
package com.hovr.devicepermissions

internal object PermissionActionResolver {
    fun resolveLocationAction(
        issue: PermissionAccessIssue,
        permissionStatus: PermissionStatus,
    ): PermissionCoordinatorAction {
        return when (issue) {
            PermissionAccessIssue.GRANTED -> PermissionCoordinatorAction.DISMISS
            PermissionAccessIssue.DEVICE_LOCATION_DISABLED ->
                PermissionCoordinatorAction.SHOW_DEVICE_LOCATION_DISABLED
            PermissionAccessIssue.NOT_DETERMINED -> PermissionCoordinatorAction.REQUEST_SYSTEM
            PermissionAccessIssue.APP_LOCATION_DENIED -> when (permissionStatus) {
                PermissionStatus.NOT_DETERMINED,
                PermissionStatus.DENIED,
                -> PermissionCoordinatorAction.REQUEST_SYSTEM
                else -> PermissionCoordinatorAction.SHOW_REQUIRED
            }
            PermissionAccessIssue.APP_NOTIFICATION_DENIED -> PermissionCoordinatorAction.DISMISS
        }
    }

    fun resolveRuntimePermissionAction(status: PermissionStatus): PermissionCoordinatorAction {
        return when (status) {
            PermissionStatus.GRANTED -> PermissionCoordinatorAction.DISMISS
            PermissionStatus.NOT_DETERMINED,
            PermissionStatus.DENIED,
            -> PermissionCoordinatorAction.REQUEST_SYSTEM
            PermissionStatus.DENIED_PERMANENTLY -> PermissionCoordinatorAction.SHOW_REQUIRED
            else -> PermissionCoordinatorAction.SHOW_REQUIRED
        }
    }
}
